File: clojure/nucleotide-count/nucleotide_count.clj
(ns nucleotide-count)

(def nucleotides "ATCG")

(defn count [nucleotide dna]
  (cond
    (some #{nucleotide} nucleotides) (clojure.core/count (filter #(= nucleotide %) dna))
    :else (throw (Exception. "invalid nucleotide"))))

(defn nucleotide-counts [dna]
  (zipmap nucleotides (map #(nucleotide-count/count % dna) nucleotides)))
  

File: ruby/nucleotide-count/nucleotide_count.rb
class Nucleotide
  attr_reader :histogram

  def self.from_dna(dna)
    Nucleotide.new(dna)
  end

  def initialize(dna)
    @histogram = {
      "A" => 0,
      "T" => 0,
      "C" => 0,
      "G" => 0
    }
    dna.each_char do |c|
      raise ArgumentError unless @histogram.key?(c)
      @histogram[c] += 1
    end
  end

  def count(nucleotide)
    @histogram[nucleotide]
  end
end
